src/Behat/Testwork/EventDispatcher/TestworkEventDispatcher.php

```diff
@@ -17,8 +17,27 @@
  *
  * @author Konstantin Kudryashov <ever.zet@gmail.com>
  */
+$isSymfony5AndPhpGreater72 = function () {
+    $reflection  = new \ReflectionClass(\Symfony\Component\EventDispatcher\EventDispatcher::class);
+    $dispatch    = $reflection->getMethod('dispatch');
+    $first_param = $dispatch->getParameters()[0];
+    switch ($first_param->getName()) {
+        case 'event':
+            // This is the new Symfony 5 event dispatcher interface
+            // This test retained from existing code. In reality it's unclear that it could ever be false, since
+            // symfony/event-dispatcher 5.x requires php: ^7.2.5 in their composer, and before that the class has the
+            // old signature.
+            return PHP_VERSION_ID > 70200;
 
-if (class_exists(\Symfony\Contracts\EventDispatcher\Event::class) && PHP_VERSION_ID > 70200) {
+        case 'eventName':
+            return false;
+
+        default:
+            throw new \UnexpectedValueException('Could not identify which version of symfony/event-dispatcher is in use, could not define a compatible TestworkEventDispatcher');
+    }
+};
+
+if ($isSymfony5AndPhpGreater72()) {
```

Comment:
Instead of checking method existance, what do you think about using by default old implementation? 
```suggestion
if (class_exists(\Symfony\Component\EventDispatcher\EventDispatcher::class) && PHP_VERSION_ID > 70200) {
```

Reply:
I'm not sure how that would help? `\Symfony\Component\EventDispatcher\EventDispatcher` exists in all versions of symfony - the class name hasn't changed, just the signature of `dispatch`

Reply:
Which is an annoying 'backward compatible' change

Taking a step back, why do we need two implementations of the class? can't we extend in a way that copes with either?

Reply:
My bad, wrong class checked, as I mistaken Event and EventDispatcher.

Reply:
@ciaranmcnulty I don't think so for two reasons:

a) symfony < and > 5 have hard typehints on dispatch that are not compatible, and you extend and replace the inherited dispatch method (I've not looked into why that is). There's no way to define your own `dispatch` with a signature that's compatible with both symfony versions.

b) even if you could somehow do that, symfony5 uses php 7.2 hints (`string` and ` : object`) that will generate parse errors if running behat on php5 which I assume is why there's the odd magic of defining that method in a trait to prevent the file being loaded in an older php environment.

Worth noting the symfony change isn't intended to be backward compatible, the method signature changed in 5.0. The issue is behat wants to attempt to support symfony 2.x - 5.x all from a single release of the behat codebase. 

It would probably have been better to make symfony5 support a breaking change and optional upgrade (and php7-only at the same time) on behat - there's a very large amount of magic/variation/duplication going on in #1256 to support both versions at once.

Reply:
Thanks for the explanation, it makes a lot of sense

OK, to make this cleaner how about we do this?

- Move the 2 implementations into separate autoloadable files (maybe remove the trait if we can) called something like `-Legacy` and `-Symfony5`
- Make this file do the `if` and then a `class_alias` to whichever implementation should be loaded
- have a dummy implementation inside an `if(false)` here to fool the composer autoloader